File: src/main/java/engineer/thomas_werner/euler/math/CollatzSequence.java
package engineer.thomas_werner.euler.math;

import java.util.function.Supplier;

public class CollatzSequence implements Supplier<Long> {

    private long value;

    public CollatzSequence(final long startValue) {
        value = startValue;
    }

    public long nextValue() {
        return value = (0 == value % 2) ? value / 2 : 3 * value +1;
    }

    @Override
    public Long get() {
        return nextValue();
    }

    public static int getLength(final long startValue) {
        if(1 == startValue)
            return 1;

        int result = 2;
        final CollatzSequence sequence = new CollatzSequence(startValue);
        while(1 != sequence.nextValue())
            result++;

        System.out.println(startValue + " (" + result + ")");
        return result;
    }

}

File: src/main/kotlin/de/huddeldaddel/euler/math/CollatzSequence.kt
package de.huddeldaddel.euler.math

import java.util.function.Supplier

class CollatzSequence(private var value: Long) : Supplier<Long> {

    fun nextValue(): Long {
        value = if (0L == value % 2) value / 2 else 3 * value + 1
        return value
    }

    override fun get(): Long {
        return nextValue()
    }

    companion object {

        fun getLength(startValue: Long): Int {
            if (1L == startValue)
                return 1

            var result = 2
            val sequence = CollatzSequence(startValue)
            while (1L != sequence.nextValue())
                result++

            println("$startValue ($result)")
            return result
        }
    }

}
